[bootloader] Add Ubaboot support

rastersoft opened the pull request:
This MR adds full Ubaboot support, both for creating firmware for devices having it, and for flashing it from the QMK tool.

It supports both the default python flasher, and the new C flasher.

Ubaboot repository: https://github.com/rrevans/ubaboot

<!--- Describe your changes in detail here. -->

It adds Ubaboot support, implementing:

- build firmware that takes into account the size used by it (512 bytes)
- flash the firmware using the QMK tool
- reboot the CPU using QK_REBOOT
- enter the programming mode using QK_BOOTLOADER

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [X] New feature
- [X] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [X] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

elpekenin commented on docs/flashing.md:
```diff
@@ -442,3 +442,23 @@ CLI Flashing sequence:
+
+## Ubaboot
+
+This is a little bootloader (less than 512 bytes) developed by RREvans. The `rules.ml` setting for this
```
```suggestion
This is a little bootloader (less than 512 bytes) developed by RREvans. The `rules.mk` setting for this
```

elpekenin commented on docs/isp_flashing_guide.md:
```diff
@@ -182,6 +182,13 @@ Precompiled `.hex` files are generally not available, but you can compile it you
 
+### Ubaboot
+
+This bootloader is designed to occupy the bare minimum possible, using less that 512 bytes of flash. It runs on ATmega32U MCUs. It can be flashed with any ISP supported by avrdude.
```
32U***4*** i guess, Zadig list also stated "32U", does it really appear with such name?

rastersoft commented:
@drashna my fault... Thanks for fixing the base branch. I did a rebase right now.

drashna commented on docs/driver_installation_zadig.md:
```diff
@@ -97,3 +97,4 @@ The device name here is the name that appears in Zadig, and may not be what the
 |`kiibohd`     |Kiibohd DFU Bootloader        |`1C11:B007`   |WinUSB |
 |`stm32duino`  |Maple 003                     |`1EAF:0003`   |WinUSB |
 |`qmk-hid`     |(keyboard name) Bootloader    |`03EB:2067`   |HidUsb |
+|`ubaboot`     |ATmega32U                     |`1d50:611c`   |libusb0|
```
```suggestion
|`ubaboot`     |ATmega32u4                    |`1d50:611c`   |libusb0|
```

drashna reviewed (commented):
ideally, flashing script should be added too.  ubaboot.py could probably be thrown in `/util/`.  Also, pyusb is installed as part of the qmk firmware stuff, so all of the dependancies to use it would appear to be preset.

rastersoft commented:
@drashna I added ubaboot.py into util, and modified the calling script to search first for system-wide installed flashers, and if it can't find them, then use the one in util.

rastersoft commented:
Or should I add code in the "install" scripts to get it (and the C version) and install (and compile if needed) them from the repos?

drashna commented:
> Or should I add code in the "install" scripts to get it (and the C version) and install (and compile if needed) them from the repos?

just in utils is okay, I think.  The Install dir is more for installing packages.  

Also, you may want to add the VID/PID combo to the udev rules

rastersoft commented:
@drashna Added VID/PID values.

sigprof commented:
The udev rule needs to be added to `util/udev/50-qmk.rules`, otherwise `qmk doctor` will complain that it's missing even after its instructions are followed (`qmk doctor` checks whether all IDs specified in `BOOTLOADER_VIDS_PIDS` in `lib/python/qmk/constants.py` are actually present in some udev rule file on the system).

Also you may extend `lib/python/qmk/flashers.py` to deal with this bootloader (that code is used by `qmk flash compiled_firmware.hex`).

rastersoft commented:
@sigprof Ops... thanks! I'll fix it now.

rastersoft commented:
@sigprof Done.

zvecr commented on util/ubaboot.py:
```diff
@@ -0,0 +1,303 @@
+        print('Rebooting')
+
+
+class Gui:
```
We would want to keep the changes as minimal as possible to mitigate maintenance issues. As such, I dont think we would want to pick up this `Gui` section of the code and its additional handling above.

rastersoft replied:
It makes sense. Removed.

About the upstream, I'll sent it too.

rastersoft replied:
All done.

zvecr commented on util/ubaboot.py:
```diff
@@ -0,0 +1,242 @@
+
+    reboot = subparsers.add_parser('reboot', help='reboot device')
+    reboot.set_defaults(mode='reboot', reboot=True)
+
+    arguments = parser.parse_args()
+    return arguments
```
```suggestion
    return parser.parse_args()
```

zvecr commented on lib/python/qmk/flashers.py:
```diff
@@ -196,6 +196,10 @@ def _flash_uf2(file):
 
 
+def _flash_ubaboot(file):
+    cli.run(['util/ubaboot.py', '--wait', 'write', file], capture_output=False)
```
Probably dont need the wait here, as the cli flashing process would have already searched for the vid:pid
```suggestion
    cli.run(['util/ubaboot.py', 'write', file], capture_output=False)
```

rastersoft replied:
But that only happens when you use `qmk flash FILE.HEX`; if you run `qmk flash -kb XXXXX -km YYYYY`, it doesn't check if the device is ready.

rastersoft replied:
(or maybe it would be better to add a patch for that case too...?)

zvecr replied:
> it doesn't check if the device is ready.

That goes through a different pathway to `flash.mk`, so you are misunderstood how this works.

rastersoft replied:
What I mean is that maybe it would be a good idea to add the same generic device check in that other pathway too, and completely remove the waiting code from ubaboot.py, thus leaving the ubaboot.py code as it's currently in upstream. It also would benefit all other programmers.

rastersoft replied:
Ok; if the idea is to move everything to CLI, then it makes no sense in adding the check now to the flash.mk pathway.

zvecr commented on util/ubaboot.py:
```diff
@@ -0,0 +1,242 @@
+#!/usr/bin/env python3
+
```
Rather than pick up the various whitespace formatting changes, its probably better to revert those and just ignore
```suggestion
# yapf: disable
```

rastersoft replied:
Done, but... what do you mean with "revert those whitespace formatting changes"? Are there changes in the use of spaces/tabs?

zvecr replied:
Diff the contents with upstream, its fairly obvious.

zvecr replied:
This shouldnt be marked as resolved when the requested actions have not been completed.

rastersoft replied:
Oh, ok... NOW I understand what you mean. I'll put the original file from the ubaboot repo (with the "retry" patch, of course).

rastersoft replied:
Ok, I think that now it's fixed.

zvecr replied:
You still want the suggestion [here](https://github.com/qmk/qmk_firmware/pull/20351#discussion_r1161344778), so that CI ignores the file.

rastersoft replied:
Agh! I probably overwrote the change accidentally! Sorry!

rastersoft commented:
good news: RRevans accepted the --wait patch in upstream. Now the python script is the same.

rastersoft commented:
@zvecr It seems that # yapf: disable only works if put after the comment block. Now it seems to avoid re-formatting.

rastersoft commented:
This should be ready...

fauxpark commented on docs/driver_installation_zadig.md:
```diff
@@ -97,3 +97,4 @@ The device name here is the name that appears in Zadig, and may not be what the
 |`kiibohd`     |Kiibohd DFU Bootloader        |`1C11:B007`   |WinUSB |
 |`stm32duino`  |Maple 003                     |`1EAF:0003`   |WinUSB |
 |`qmk-hid`     |(keyboard name) Bootloader    |`03EB:2067`   |HidUsb |
+|`ubaboot`     |ATmega32u4                    |`1d50:611c`   |libusb0|
```
```suggestion
|`ubaboot`     |*none*                        |`1D50:611C`   |libusb0|
```

rastersoft commented:
Any news...?
